Sources/StreamChat/Workers/MessageUpdater.swift

```diff
@@ -488,26 +495,36 @@ class MessageUpdater: Worker {
     ///   - messageId: The message identifier.
     ///   - completion: The completion. Will be called with an error if smth goes wrong, otherwise - will be called with `nil`.
     func unpinMessage(messageId: MessageId, completion: ((Error?) -> Void)? = nil) {
+        var pinning: MessagePinning = .noExpiration
+        
         database.write({ session in
             guard let messageDTO = session.message(id: messageId) else {
                 throw ClientError.MessageDoesNotExist(messageId: messageId)
             }
 
-            switch messageDTO.localMessageState {
-            case nil, .pendingSync, .syncingFailed, .deletingFailed:
-                session.unpin(message: messageDTO)
-                messageDTO.localMessageState = .pendingSync
-            case .pendingSend, .sendingFailed:
-                session.unpin(message: messageDTO)
-                messageDTO.localMessageState = .pendingSend
-            case .sending, .syncing, .deleting:
-                throw ClientError.MessageEditing(
-                    messageId: messageId,
-                    reason: "message is in `\(messageDTO.localMessageState!)` state"
-                )
+            pinning = .init(expirationDate: messageDTO.pinExpires?.bridgeDate)
+            session.unpin(message: messageDTO)
+        }, completion: { [weak self, weak repository] error in
+            if let error {
+                completion?(error)
+                return
+            }
+            
+            let endpoint: Endpoint<EmptyResponse> = .pinMessage(
+                messageId: messageId,
+                request: .init(set: .init(pinned: false))
+            )
+            
+            self?.apiClient.request(endpoint: endpoint) { result in
+                switch result {
+                case .success:
+                    completion?(nil)
+                case .failure(let apiError):
+                    repository?.undoMessageUnpinning(on: messageId, pinning: pinning) { _ in
+                        completion?(apiError)
+                    }
```

Comment:
Feels like it would be clearer if we just have a single unpin and pin method. Therefore, I would like to propose removing the `MessageRepository` change and having a private unpin and pin here which only changes database. e.g. something like `pinLocalMessage`. Then we can use that within internal method here which:
1. modifies database state (through `unpinLocalMessage`)
2. makes an api request
3. reverts the local change if the API request fails (`pinLocalMessage` with the existing pinning state)

WDYT?

Reply:
Could you check out the last commit? I think I got your point, but I might have missed something.